Makefile

```diff
@@ -19,50 +26,48 @@ $(error "error Unexpected OS; Only Linux or Darwin supported.")
 endif
 endif
 
-# Environment
-SHELL := /bin/bash
-BASE_PATH := $(shell pwd)
-PUBLIC_PATH := $(BASE_PATH)/public
-THEMES_PATH := $(BASE_PATH)/themes
-STATIC_PATH := $(BASE_PATH)/static
-THEME_NAME := $(shell basename $(HUGO_THEME))
-THEME_PATH := $(THEMES_PATH)/$(THEME_NAME)
-HUGO_PATH := $(BASE_PATH)/.hugo
-HUGO_URL = https://github.com/spf13/hugo/releases/download/v$(HUGO_VERSION)/$(HUGO_TAR_FILE_NAME)
-HUGO_NAME := hugo_$(HUGO_VERSION)_$(ARCH)_$(OS)
-HUGO_URL_NAME := hugo_$(HUGO_VERSION)_$(URL_ARCH)-$(URL_OS)
-
 # Tools
-CURL = curl -L
-HUGO = $(HUGO_PATH)/hugo
-MKDIR = mkdir -p
-GIT = git
+CURL := curl -L
+HUGO := $(HUGO_PATH)/hugo
+MKDIR := mkdir -p
+JS_PACKAGE_MANAGER := yarn
 UNCOMPRESS := tar -zxf
 
 # Rules
 all: build
 
-init:
-	@if [ "$(HUGO_THEME)" == "" ]; then \
-		echo "ERROR! Please set the env variable 'HUGO_THEME' (http://mcuadros.github.io/autohugo/documentation/working-with-autohugo/)"; \
-		exit 1; \
-	fi;
+# Ensures the Hugo binary existance
+$(HUGO): $(HUGO_TAR_PATH)
+	$(UNCOMPRESS) $(HUGO_TAR_PATH) --directory=$(HUGO_PATH);
+
+# Downloads the hugo binary
+$(HUGO_TAR_PATH):
+	echo "Downloading $(HUGO_URL)"
+	$(MKDIR) $(HUGO_PATH)
+	$(CURL) $(HUGO_URL) -o $(HUGO_TAR_PATH)
 
-dependencies: init
-	@if [[ ! -f $(HUGO) ]]; then \
-		$(MKDIR) $(HUGO_PATH); \
-		cd $(HUGO_PATH); \
-		echo "Downloading $(HUGO_URL)"; \
-		$(CURL) $(HUGO_URL) -o $(HUGO_TAR_FILE_NAME); \
-		$(UNCOMPRESS) $(HUGO_TAR_FILE_NAME); \
-	fi;
+# Ensures hugo dependencies
+hugo-dependencies: $(HUGO_TAR_PATH) $(HUGO)
 
-build: dependencies
-	$(HUGO) -t $(THEME_NAME) --baseURL $(BASE_URL)
+# Prepares yarn
+js-dependencies:
+	$(JS_PACKAGE_MANAGER) install --force
+	$(JS_PACKAGE_MANAGER) build
+
+# Prepares project dependencies
+project-dependencies: hugo-dependencies js-dependencies
+
+# Builds hugo
+hugo-build:
+ifeq ($(ENVIRONMENT),staging)
+	$(HUGO) --baseURL=$(BASE_URL) --theme=$(HUGO_THEME_NAME) --buildDrafts --buildFuture
+else
+	$(HUGO) --baseURL=$(BASE_URL) --theme=$(HUGO_THEME_NAME)
+endif
 
-server: dependencies
-	$(HUGO) server -t $(THEME_NAME) -D -w --baseURL $(BASE_URL)
+## Builds project
+build: project-dependencies hugo-build
 
-clean:
-	rm -rf $(HUGO_PATH)
-	rm -rf $(THEME_PATH)
+# Runs hugo server
+hugo-server:
```

Comment:
Why this is not called server anymore? Why it doesn't depend on `dependencies`?

Reply:
The old `make serve` didn't fit our necesities so it was dropped to be implemented via a new PR. It was not done in this PR because it would require to add a lot of stuff needed by the `make serve` rule, that will use `yarn` and `terminusMaximus`

The new `make serve` rule will call `project-dependencies`.

You can see the implementation of the `make serve` rule in https://github.com/src-d/blog/pull/148/commits/bc92d753e00ba47e9dbbf77c8d364dfa021fa703 from PR https://github.com/src-d/blog/pull/148
These are the key parts of that feature:
- `make serve` rule: https://github.com/src-d/blog/commit/bc92d753e00ba47e9dbbf77c8d364dfa021fa703#diff-b67911656ef5d18c4ae36cb6741b7965R72
- `yarn start` rule: https://github.com/src-d/blog/commit/bc92d753e00ba47e9dbbf77c8d364dfa021fa703#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R8
- `terminusMaximus` configuration: https://github.com/src-d/blog/commit/bc92d753e00ba47e9dbbf77c8d364dfa021fa703#diff-6565c975017624099a7d09dbd14ee76bR1

Reply:
The old `make serve` didn't fit our necesities so it was dropped to be implemented via a new PR. It was not done in this PR because it would require to add a lot of stuff needed by the `make serve` rule, that will use `yarn` and `terminusMaximus`

The new `make serve` rule will call `project-dependencies`.

You can see the implementation of the `make serve` rule in https://github.com/src-d/blog/pull/148/commits/bc92d753e00ba47e9dbbf77c8d364dfa021fa703 from PR https://github.com/src-d/blog/pull/148
These are the key parts of that feature:
- `make serve` rule: https://github.com/src-d/blog/commit/bc92d753e00ba47e9dbbf77c8d364dfa021fa703#diff-b67911656ef5d18c4ae36cb6741b7965R72
- `yarn start` rule: https://github.com/src-d/blog/commit/bc92d753e00ba47e9dbbf77c8d364dfa021fa703#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R8
- `terminusMaximus` configuration: https://github.com/src-d/blog/commit/bc92d753e00ba47e9dbbf77c8d364dfa021fa703#diff-6565c975017624099a7d09dbd14ee76bR1

Reply:
I guess it should be called then `make server`, isn't it?

Anyway, I guess this new action `hugo-build` will also need the dependencies installed, isn't it? If it's not good anymore, then I'd say we should remove it...

Reply:
Agreed with @dpordomingo that this should be part of upcoming documentation